File: Parallel Programming - Weeks 4-7/Week 5/2 - Join and Sleep/JoinExample.java
/*
 * Two threads running at the same time. Each one prints three
 * lines with a one-second pause between them, and the main
 * thread waits for both with join.
 *
 * Run it and watch the two threads interleave their output. The
 * exact order depends on the scheduler. Compare with
 * join_example.c in the same folder.
 */
public class JoinExample {

    public static void main(String[] args) throws InterruptedException {

        // First worker. Lambda over a Runnable.
        Thread a = new Thread(() -> {
            for (int i = 1; i <= 3; i++) {
                System.out.println("Thread A step " + i);
                try { Thread.sleep(1000); } catch (InterruptedException e) {}
            }
        });

        // Second worker. Same shape, different name.
        Thread b = new Thread(() -> {
            for (int i = 1; i <= 3; i++) {
                System.out.println("Thread B step " + i);
                try { Thread.sleep(1000); } catch (InterruptedException e) {}
            }
        });

        a.start();
        b.start();

        // Wait for both before we finish.
        a.join();
        b.join();

        System.out.println("Both threads are done.");
    }
}

File: Parallel Programming - Weeks 4-7/Week 5/2 - Join and Sleep/join_example.c
/*
 * Same as JoinExample.java, rewritten with pthreads.
 *
 * A few things worth comparing with the Java side:
 *
 *   - Java's Thread.sleep takes milliseconds. C's sleep takes
 *     whole seconds. For sub-second pauses in C you'd use
 *     usleep (microseconds) or nanosleep.
 *
 *   - The work of each thread lives in a named function rather
 *     than an inline lambda. There are no lambdas in plain C.
 *
 *   - We can use one function for both threads and pass the name
 *     through the void* argument. Java's two lambdas could also
 *     have been collapsed into one Runnable with a constructor
 *     argument, but usually you just write two.
 */

#include <stdio.h>
#include <unistd.h>   /* sleep */
#include <pthread.h>

void* worker(void* arg) {
    const char* name = (const char*) arg;
    for (int i = 1; i <= 3; i++) {
        printf("Thread %s step %d\n", name, i);
        sleep(1);
    }
    return NULL;
}

int main() {
    pthread_t a, b;

    /* Pass a string literal through the void* slot. Inside the
       worker we cast it back to const char*. */
    pthread_create(&a, NULL, worker, "A");
    pthread_create(&b, NULL, worker, "B");

    pthread_join(a, NULL);
    pthread_join(b, NULL);

    printf("Both threads are done.\n");
    return 0;
}
